src/Microsoft.VisualStudio.Threading/AsyncReaderWriterResourceLock`2.cs

```diff
@@ -768,6 +769,7 @@ private Task PrepareResourceAsync(TResource resource, CancellationToken cancella
             AsyncReaderWriterResourceLock<TMoniker, TResource>.Helper.ResourceState finalState = forConcurrentUse ? ResourceState.Concurrent : ResourceState.Exclusive;
 
             Task? preparationTask = null;
+            TaskScheduler taskScheduler = TaskScheduler.Current.MaximumConcurrencyLevel > 1 ? TaskScheduler.Current : TaskScheduler.Default;
```

Comment:
This makes me very nervous. There are other TaskSchedulers that may be in the system besides the ones you're thinking of. ReSharper for example has their own custom schedulers. I don't know how they set their `MaximumConcurrencyLevel` property, but in the past, folks who have used `TaskScheduler.Current` (usually by accident) often find themselves in deadlocks when their continuations inadvertently rely on a Resharper scheduler which has unknown and unwanted semantics.

Reply:
yes, I am not fan of this logic, but have trouble to provide an alternative solution. Another possibility is to create a new virtual method to let the lock to potentially provide the TaskScheduler for PrepareResource work, and this can leave using TaskScheduler.Default, but the project lock service can choose the set of TaskScheduler to be compatible here. Of course, this would introduce a hard-to-explain contract. Let me know if you think that is a better solution here.

Reply:
Well, that sounds safer. And I'm a fan of [as simple as possible, but no simpler](https://www.bing.com/search?q=as+simple+as+possible%2C+but+no+simpler&cvid=7bf666f97b10476c803d00d3678aa649&aqs=edge..69i57j0l5j69i11004.333j0j1&pglt=43&FORM=ANNAB1&PC=U531). So in lieu of a simpler idea that is still safe, I like your virtual method idea. I'm guessing your override will do an actual concrete type check on the TaskScheduler to make sure it's the one you have in mind as safe and performant.

Reply:
sounds good, and it could make it much easier to apply a throttling of evaluations by using a special scheduler, instead of TaskScheduler.Default if we want to, and use a priority one for UI block/critical responsiveness path there if necessary

Reply:
I updated the PR, and please take a look.